### PR TITLE
feat: track finding outcomes on PR merge

### PR DESCRIPTION
**Pull request by bb-amir**
## Summary

- Track what happens to each Baloo finding after a PR is merged: was it actioned, disputed, acknowledged, or ignored?
- On PR merge webhook, label each finding with an outcome based on code changes near the flagged line, thread resolution, and developer reply sentiment
- New `/dashboard/outcomes` page with hit rate, noise rate, severity/category breakdowns, weekly trends, and outcome distribution charts
- New `finding_outcomes` DB table with Alembic migration

## Motivation

Build the measurement layer for review quality. Knowing which findings lead to code changes vs. get ignored lets us improve prompts and filtering over time. Global metrics first, per-repo and per-contributor optimization later.

## Changes

- **Model + Migration**: `FindingOutcome` model, Alembic migration 004
- **Signal Collection**: `baloo/outcomes/signals.py` — sentiment classification, code change detection, thread signal extraction
- **Labeler**: `baloo/outcomes/labeler.py` — orchestrates signal collection, applies priority logic (actioned > disputed > acknowledged > ignored), persists outcomes
- **Webhook**: Handles `pull_request:closed` + `merged=true` to trigger outcome labeling
- **Dashboard**: Queries in `DashboardService`, new `/dashboard/outcomes` route + template with Chart.js visualizations
- **Backfill**: Separate `tool/backfill-outcomes` branch with one-time script (not included in this PR)

## Test plan

- [ ] 376 tests passing (55 new tests across outcomes and dashboard)
- [ ] Lint and format clean (`ruff check`, `black --check`)
- [ ] Verify `/dashboard/outcomes` renders with mock data
- [ ] Verify merge webhook triggers `label_pr_outcomes`
- [ ] Verify idempotency (labeling same PR twice doesn't duplicate rows)

### CODE REVIEW

File: baloo/outcomes/labeler.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            }
+        else:
+            child_comments.setdefault(reply_to, []).append(comment_dict)
+
+    # Attach children to their root threads
+    for root_id, thread in root_comments.items():
+        if root_id in child_comments:
+            thread["comments"].extend(child_comments[root_id])
+
+    threads = list(root_comments.values())
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/labeler.py:111]: Fixed. Wrapped the entire function body in try/except with structured logging.

File: baloo/github/webhook_handler.py [hunk 2]
```diff
@@ -381,6 +382,17 @@ async def handle_webhook(
                 background_tasks.add_task(lambda: None)
 
                 return {"status": "queued", "queue_depth": waiting}
+            elif action == "closed":
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/github/webhook_handler.py:385]: Fixed. Task is now stored and has a done_callback that logs exceptions.

File: baloo/outcomes/labeler.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    async with session_factory() as session:
+        # Find all findings for this PR
+        stmt = (
+            select(Finding)
+            .join(Review, Finding.review_id == Review.id)
+            .where(Review.repo_full_name == repo_full_name, Review.pr_number == pr_number)
+        )
+        result = await session.execute(stmt)
+        findings = result.scalars().all()
+
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/labeler.py]: The unique index on finding_id is the real safety net here. With the error handling from finding 1 now in place, an IntegrityError from a race is caught and logged rather than silently swallowed. A full upsert adds complexity for a scenario (duplicate webhook delivery for same PR merge) that is rare in practice.

File: baloo/outcomes/signals.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    for raw_line in diff.splitlines():
+        # File header
+        m = file_header_re.match(raw_line)
+        if m:
+            current_file = m.group(1)
+            in_target_file = current_file == file_path
+            new_line = 0
+            continue
+
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/signals.py:57]: Fixed. Added explicit guard for lines starting with backslash before the context-line else branch.

File: baloo/outcomes/labeler.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                len(findings),
+                repo_full_name,
+                pr_number,
+            )
+            return
+
+    # Fetch signals from GitHub (outside DB session)
+    diff_text, threads = await fetch_merge_signals(repo_full_name, pr_number, installation_id)
+
+    # Label each finding
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/labeler.py]: Fixed. Findings are now snapshotted as plain dicts (id, review_id, file_path, line_number) before the session closes.

File: baloo/dashboard/queries.py [hunk 6]
```diff
@@ -383,3 +383,165 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+                    filters.append(FindingOutcome.repo_full_name == repo_filter)
+                return filters
+
+            # --- Totals and rates ---
+            total = (
+                await session.execute(select(func.count(FindingOutcome.id)).where(*_base_filters()))
+            ).scalar() or 0
+
+            outcome_rows = (
+                await session.execute(
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:408]: By design per spec — acknowledged is a distinct category (positive reply, no code change). The outcome breakdown card already shows all four counts explicitly so the gap is visible. Changing the metric definitions would conflate categories we want to track separately.

File: baloo/dashboard/templates/outcomes.html [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      <span class="inline-flex items-center px-2 py-0.5 rounded text-xs font-medium bg-green-100 text-green-800">actioned: {{ outcomes.actioned }}</span>
+      <span class="inline-flex items-center px-2 py-0.5 rounded text-xs font-medium bg-blue-100 text-blue-800">acknowledged: {{ outcomes.acknowledged }}</span>
+      <span class="inline-flex items-center px-2 py-0.5 rounded text-xs font-medium bg-red-100 text-red-800">disputed: {{ outcomes.disputed }}</span>
+      <span class="inline-flex items-center px-2 py-0.5 rounded text-xs font-medium bg-gray-100 text-gray-800">ignored: {{ outcomes.ignored }}</span>
+    </div>
+  </div>
+</div>
+
+<!-- Charts grid -->
+<div class="grid grid-cols-1 lg:grid-cols-2 gap-6 mb-6">
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/templates/outcomes.html:63]: Fixed. Using .get(key, 0) for all four outcome counts.

File: baloo/dashboard/queries.py [hunk 8]
```diff
@@ -383,3 +383,165 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+                        func.count(FindingOutcome.id),
+                    )
+                    .join(Finding, FindingOutcome.finding_id == Finding.id)
+                    .where(*_base_filters())
+                    .group_by(Finding.category, FindingOutcome.outcome)
+                )
+            ).all()
+
+            cat_map: dict[str, dict] = {}
+            for cat, outcome, cnt in category_rows:
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:468]: SQLite is only used in tests, Postgres in production. The year-boundary edge case affects at most 1 week of data per year in the test path. Adding %G-%V requires SQLite 3.38+ which constrains CI environments for minimal gain. Leaving as-is.

File: baloo/github/webhook_handler.py [hunk 9]
```diff
@@ -381,6 +382,24 @@ async def handle_webhook(
...
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/github/webhook_handler.py:385]: All 8 threads were replied to before pushing — check the thread history. The threads weren't marked as resolved on GitHub, but each has an explicit response.

File: baloo/outcomes/labeler.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        pr_url = f"{client.base_url}/repos/{repo_full_name}/pulls/{pr_number}"
+        diff_resp = await http.get(
+            pr_url,
+            headers={**headers, "Accept": "application/vnd.github.v3.diff"},
+        )
+        diff_resp.raise_for_status()
+        diff_text = diff_resp.text
+
+        # Fetch review comments (paginated)
+        comments_url = f"{client.base_url}/repos/{repo_full_name}/pulls/{pr_number}/comments"
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/labeler.py:73]: Fixed. fetch_merge_signals now handles 406 by falling back to empty diff (code-change detection skipped, thread signals still collected).

File: baloo/dashboard/queries.py [hunk 11]
```diff
@@ -383,3 +383,165 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+            outcomes = {r[0]: r[1] for r in outcome_rows}
+
+            actioned = outcomes.get("actioned", 0)
+            disputed = outcomes.get("disputed", 0)
+            ignored = outcomes.get("ignored", 0)
+
+            hit_rate = round(actioned / total * 100, 1) if total else 0.0
+            noise_rate = round((disputed + ignored) / total * 100, 1) if total else 0.0
+
+            # --- By severity ---
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:423]: Declined in previous review — acknowledged is intentionally a distinct category per spec. The outcome breakdown card shows all four counts explicitly.

File: baloo/dashboard/queries.py [hunk 12]
```diff
@@ -383,3 +383,165 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+            }
+
+            # --- By category ---
+            category_rows = (
+                await session.execute(
+                    select(
+                        Finding.category,
+                        FindingOutcome.outcome,
+                        func.count(FindingOutcome.id),
+                    )
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:460]: Declined in previous review — SQLite is test-only, Postgres in production. Year-boundary edge case is low impact.

File: baloo/outcomes/labeler.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+                repo_full_name,
+                pr_number,
+            )
+            return
+
+        # Fetch signals from GitHub (outside DB session)
+        diff_text, threads = await fetch_merge_signals(repo_full_name, pr_number, installation_id)
+
+        # Label each finding
+        async with session_factory() as session:
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/outcomes/labeler.py:176]: Fixed. Now uses closest-match (smallest distance) instead of first-match within the ±5 window.

File: baloo/dashboard/queries.py [hunk 14]
```diff
@@ -383,3 +383,165 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+            cat_map: dict[str, dict] = {}
+            for cat, outcome, cnt in category_rows:
+                if cat not in cat_map:
+                    cat_map[cat] = {"total": 0, "actioned": 0}
+                cat_map[cat]["total"] += cnt
+                if outcome == "actioned":
+                    cat_map[cat]["actioned"] += cnt
+            category_data = {
+                cat: {
+                    "total": v["total"],
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:476]: Fixed. Weekly trends map now initializes with all four outcome keys including acknowledged.

File: baloo/dashboard/queries.py [hunk 15]
```diff
@@ -383,3 +383,171 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+            severity_data = {
+                sev: {
+                    "total": v["total"],
+                    "actioned": v["actioned"],
+                    "hit_rate": round(v["actioned"] / v["total"] * 100, 1) if v["total"] else 0.0,
+                }
+                for sev, v in sev_map.items()
+            }
+
+            # --- By category ---
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:453]: Declined — same reasoning as prior reviews. Test-only path, low impact.

File: baloo/dashboard/queries.py [hunk 16]
```diff
@@ -383,3 +383,171 @@ async def get_analytics_data(days: int = 30) -> dict:
...
+                if outcome == "actioned":
+                    cat_map[cat]["actioned"] += cnt
+            category_data = {
+                cat: {
+                    "total": v["total"],
+                    "actioned": v["actioned"],
+                    "hit_rate": round(v["actioned"] / v["total"] * 100, 1) if v["total"] else 0.0,
+                }
+                for cat, v in cat_map.items()
+            }
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/queries.py:481]: By design — hit_rate and noise_rate are defined per spec. The acknowledged raw counts are now tracked in weekly data for future use.

File: baloo/dashboard/templates/outcomes.html [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+      datasets: [{
+        label: 'Hit Rate %',
+        data: catHitRates,
+        backgroundColor: '#6366f1',
+      }]
+    },
+    options: {
+      responsive: true,
+      indexAxis: 'y',
+      plugins: { legend: { display: false } },
```
[the remark this answers is not in the document]
Comment (bb-amir) [reply baloo/dashboard/templates/outcomes.html:150]: Fixed. Chart now initializes all 4 known outcome types with 0 defaults.
